relay-pii/src/processor.rs

```diff
@@ -16,9 +16,112 @@ use relay_protocol::{Annotated, Array, Meta, Remark, RemarkType, Value};
 use crate::compiledconfig::{CompiledPiiConfig, RuleRef};
 use crate::config::RuleType;
 use crate::redactions::Redaction;
-use crate::regexes::{self, ANYTHING_REGEX, PatternType, ReplaceBehavior};
+use crate::regexes::{
+    self, ANYTHING_REGEX, CREDITCARD_REGEX, EMAIL_REGEX, IBAN_REGEX, PatternType, ReplaceBehavior,
+    US_SSN_REGEX,
+};
 use crate::utils;
 
+/// A PII pattern for smart scrubbing with its associated metadata.
+struct SelectedPattern<'a> {
+    regex: &'a regex::Regex,
+    replacement: &'static str,
+    remark_suffix: &'static str,
+}
+
+/// Smart PII scrubbing for logentry.formatted that preserves context while removing sensitive
+/// data.
+fn smart_scrub_logentry_formatted(
+    value: &mut String,
+    meta: &mut Meta,
+    state: &ProcessingState<'_>,
+    compiled_config: &CompiledPiiConfig,
+) -> ProcessingResult {
+    let patterns = [
+        SelectedPattern {
+            regex: &EMAIL_REGEX,
+            replacement: "[Email]",
+            remark_suffix: "email",
+        },
+        SelectedPattern {
+            regex: &CREDITCARD_REGEX,
+            replacement: "[CreditCard]",
+            remark_suffix: "creditcard",
+        },
+        SelectedPattern {
+            regex: &IBAN_REGEX,
+            replacement: "[IBAN]",
+            remark_suffix: "iban",
+        },
+        SelectedPattern {
+            regex: &US_SSN_REGEX,
+            replacement: "[SSN]",
+            remark_suffix: "ssn",
+        },
+    ];
+
+    let original_length = value.len();
+    let mut applied_patterns = Vec::with_capacity(patterns.len());
+
+    for pattern in &patterns {
+        let new_value = pattern.regex.replace_all(value, pattern.replacement);
+        if matches!(new_value, std::borrow::Cow::Owned(_)) {
+            *value = new_value.into_owned();
+            applied_patterns.push(pattern.remark_suffix);
+        }
+    }
+
+    // Add all remarks for applied patterns and set original length if needed.
+    if !applied_patterns.is_empty() {
```

Comment:
You don't need to check the emptiness of `applied_patterns` separately; if it is empty then nothing will happen in the `for suffix …` block or the `if original_length …` block.
```suggestion
```